Srs improvement

flinkeKoralle opened the pull request:
If speed is selected while in SRS, mode reverts to OPEN CLB

### Description of Changes
If speed knob is pulled in SRS TO mode, the mode reverts to OPEN CLB if FCU altitude is above current altitude.
A triple click sounds ( not implemented yet )

### Screenshots (optional)

### Bugs fixed (if any)
<!-- If you fixed any bugs, describe them here. State issue number if applicable. -->

### Checklist:
<!-- [ ] = Unchecked, [x] = Checked. -->
* [x] My changes follow the Contributing Guidelines. <!-- See CONTRIBUTING.md to verify. -->
* [x] My changes implement realistic features. <!-- Only aircraft changes require this. -->
* [ ] Please have a main Developer test my changes before merging. <!-- We will always briefly test, but if it needs a "full" test, please check). -->

<!-- If you changes are not ready for merging, please submit this as a draft pull request. If it is ready, submit it as a regular pull request. -->

Commit:
retrieve dep.elevation and print

Commit:
before dashes

Commit:
accelAltFt

Commit:
created setAccelAlt func

Commit:
clr key on init and perf/to working

Commit:
conform tour 6.10.23

Commit:
split line for TOPERF page for thrred/accel

Commit:
Accel/ThrRed working as tested

Commit:
Merge branch 'dev' into mcdu-fmgc-merger

Commit:
corrected manual thr/acc alt with no depApt

Commit:
introducing ga-accel and ga-thrRed variables

Commit:
print accel/thrred accordingly

Commit:
ga accel/trhRed vars working- MCDU not correct

Commit:
trying to change fonrt size in perfga

Commit:
cannot change fontsize prfga page

Commit:
input in perfga almost working

Commit:
check if acc > thrRed

Commit:
round error corrected in acc/thrRed

Commit:
Startpoint for GA Property insertion

Commit:
improved imput on perf-ga page

Commit:
setting fonts

Commit:
input on perf ga further improved

Commit:
fadec thrRed changed in ga by variables

Commit:
minor errors corrected

Commit:
adding comments

Commit:
Minor change to test push request

Commit:
testing a timer with FD1 button

Commit:
default fcu alt set to 100 ft

Commit:
fcu alt selection for 1k position corrected

Commit:
finding FCU outputs

Commit:
Merge https://github.com/legoboyvdlp/A320-family into PERF-GA-Page-Accel/ThrRed-Input-correction

Commit:
Merge https://github.com/legoboyvdlp/A320-family into FCU-Preselect

Commit:
spd preselection timer in managed
todo: separate fcu selected speed and autoflight commanded speed

Commit:
preselection working
todo: spd adjust in managed gets fmgc speed instead of current

Commit:
something wrong here

Commit:
spd window indication corrected
todo: fcu default spd is 0. should be 100.

Commit:
fcu speed window corrected
todo: if FD1+2 off then revert to selected speed

Commit:
working commit. adding spd window functionality
todo: implement close/open spd window in FCU.nas file, simplifying the a320.flightdeck.xml file

Commit:
startpoint for new managed speed selection
todo: remove custom.input.spdManaged -> not needed
managed speed stops by itself if not valid

Commit:
not working to get back to selected

Commit:
interim: fms gets into selected when spd pushed
todo: check when fcu-spd is reset to selected

Commit:
fmgc logic and fcu working todo: fcu cleanup and spd window

Commit:
separation of mngd spd active and valid
todo: display correctly on fcu
show dashes and no dot when in done phase

Commit:
working on fcu indication at startup
todo: phase transition check (done phase)
and redundant variables

Commit:
start to refine FCU spd window

Commit:
selected spd on gnd inhibited

Commit:
mngSpd triggered by FDs on gnd

Commit:
mng speed triggered by v2 on gnd
TODO: mng spd triggered by to/mct

Commit:
before SRS trigger

Commit:
Managed speed on gnd as expected.
TODO: selected spd only after 5s airborne

Commit:
sel spd inhibit until 5s after lift-off
TODO: sel spd disarms on gnd at eng start
sel spd not poss. on gnd with engs running and fd on.

Commit:
Merge branch 'FCU-Preselect' into SPD-Window

Commit:
Merge branch 'PERF-GA-Page-Accel/ThrRed-Input-correction' into SPD-Window

Commit:
sel spd inhibit with one eng running

Commit:
align on gnd
TODO: spd push on gnd, if pos lost no mangd spd

Commit:
FCU SPD window behaviour improved
TODO: move Mach/Speed changeover to a property and activate via listener.

Commit:
Cleaned up Managed Speed vars. Preparing to merge last redundant vars:
fmgc.Custom.Input.spdManaged has same functionality as fmgc.FMGCNodes.mngSpdActive

Commit:
consolidated all managed speed into mngSpdActive, and spdWindowOpen

Commit:
removed unused "fmgcSpeedValid"

Commit:
SRS mode reversion when speed selected fixed
todo: it should only be possible if FCU alt > act. alt

Commit:
Speed won't be selected if FCU alt < act. alt

Commit:
Fixed logic to cover other cases

Commit:
Thrust reduction alt is changed, when GA phase is active

Commit:
Merge branch 'dev' into SRS-Improvement

Commit:
INITA crz fl input improved
TODO: 1. improve PROG input
2. correct INITA temp font size

Commit:
starting to change FMGCphase variable to a property (fmgc.FMGCNodes.phase)

Commit:
FMGC phases change. but can't arm appr phase in crz lvl.
Also mngd speed doesn't go to vapp once new crz alt is reached
SOLUTION:
introduce properties for vert mode change and setlisteners.

Commit:
phases change ok, but decel does not take place at appr phase
TODO: track down fmgc.decel variable

Commit:
Still no deceleration. managed speed is somehow constrained.

Octal450 commented:
Hi,
Circling back re the forum discussion

Is this the new request or the previous one? Just confirming for before I evaluate.

Regards,
Josh

flinkeKoralle commented:
> Hi, Circling back re the forum discussion
> 
> Is this the new request or the previous one? Just confirming for before I evaluate.
> 
> Regards, Josh

Yes this is the request that implements all the ones before that.
I am working on a better fms-phase logic. But it's not ready yet. 

I merged the latest dev branch with my latest code with no problems.
So this pull shouldn't make any problems.

Commit:
now decelerating working in appr phase
TODO: move phase transition to properties listeners.

Octal450 commented:
Hi thanks
About the fms phases, be advised that there is possibility I may need to rework this code in the future. I'm not sure yet, but also you may continue.

I will be with this request soon.

Kind Regards,
Josh

flinkeKoralle commented:
Speeds and phases are deeply interconnected.

I will continue working on it. I will open a new pull request once it is ready for testing.

Greetings,
Maxi

Commit:
change to crz phase by listener

Commit:
putting n1>85% to properties not finished
TODO: test if phase changes correctly solely on properties

Commit:
conditions from preflight to t/o as properties (n1>85 + srs + gs>90)
TODO: needds testing and removing if statement from loop

Commit:
optimizing ground listener for srs engagement conditions

Commit:
Initial point for GA SRS speed improvement:
TODO: speed jumps in GA phase

Commit:
phase changes to crz if new level inserted.
TODO: phase does not change to crz if in appr. Check pitch mode

Commit:
listeners moved because of dependency of functions.
TODO: centralize listeners. but not yet as long as everything is working

Commit:
startpoint for new SRS engagement (event based instead of ask in a loop)
   todo: remove all go around related if statements and leave one srs listener

Commit:
phase changes back to crz by reentering crz alt

Commit:
selecte speed inhibition with one engine running working again
TODO: get out of SRS if in selected speed

Commit:
inserted srsga inhibition if not in air for 5 secs
todo: check if it kicks in after 5 secs

Commit:
removing FMGC Target Knots PFD filter in fmgc-drivers.
work in progress: todo: set correct srs to speed in managedspd loop.
correct the listener for indicated speed
check if correct transition from srs/to to clb

Commit:
SRS TO shows correct speed.
Todo: SRS GA speed is wrong, phase change after ga accel alt reached
forcing a new SRS speed.

Commit:
introduced switch for takeoff accel alt in a320-libraries for fms phase switching
todo: check athr at thr.red alt vs accel alt
check if ga phase can change to clb

Commit:
srs to disengagement added
todo:
strange behaviour after toga below accel alt (went into appr phase no acceleration -> stays in srs)
test disengagement conditions
maybe rename the 400ft radio alt in FWC.

Commit:
open climb not working any more
todo: find error in vert mode 4
pfd speed target not in sync with it-autoflight/input

Commit:
restored fmgc-drivers.xml to previous state. cause of strange open climb bug
todo: srs speed at engagement wrong with flaps out of 0
indication of managed speed wrong it-autoflight/input/kts not indicated in pfd

Commit:
srsga speed stuck at first engagement
todo: reset speed at new engagement
introduce no speed change above ga-accel-spd
check if clamp is correct with oei

Commit:
SRS mode reversion if sel-spd and FCU alt below act alt
mode reverts to v/s
todo: speed bug does not move to managed speed when in SRS

Commit:
Merge branch 'dev' into GA-SRS-speed-improvement

Commit:
removed line asking for JSBsim

line invoked a nasal runtime error

Commit:
inhibit selected speed in srs

tested in level d as it is not documented in fcom
todo: magenta speed bug does not move to srs speed

Commit:
Merge branch 'dev' into GA-SRS-speed-improvement

flinkeKoralle commented:
It can be merged to dev.

flinkeKoralle commented:
It can be merged to dev.

flinkeKoralle commented:
It can be merged to dev

Commit:
GA SRS reverts to OPEN CLB above GA accel altitude

todo: 1.Speed bug still not showing right speed
2. activation of apporach phase from go aroud not possible.

Commit:
fixed bug in file ../Instruments/MCDU/MCDU.nas

Property fmgc.FMGCNodes.mngSpdActive was causing errors
Put it into a variable

Commit:
error ln5131: cannot insert any further code line

seems memory error in fn: update_common()
todo: shorten code to experiment additional lines

Commit:
moved PERFGA page to file MCDU-funcs.nas

working. startpoint for adding lines to pages

Commit:
put ACTIVATE APPR PHASE in PERFGA page

can now get out of ga-phase with activation of
appr phase
todo: check activation of crz phase is very late.
what happens if toga set at crz alt?

Commit:
listener for thrlvr2 and minor edits

for takeoff phase
todo: pfd ias not in alignment with fmgc speed

Commit:
GA speed now correct on PFD

srs speed almost done.
todo: check FMA armed modes in GA

Commit:
removed me.updateFma(). Does not exist anymore

todo: alt hold does not disengage with SRS engagement.
remove boxes from armed lines

Commit:
removed vert/lat armed boxes

they dont exist.
todo: alt should not engage if SRS conditions still met
eg. toga set
